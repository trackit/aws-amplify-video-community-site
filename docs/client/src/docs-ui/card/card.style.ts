import {css} from "emotion";
import {MQFablet, MQDesktop} from "../../amplify-ui/styles/media";

export const graphicStyle = css`
  margin-right: 0.75rem;
  width: 3rem;
  min-width: 3rem;

  &.vertical {
    margin-right: 0;
    text-align: center;
    margin-bottom: 0.5rem;
  }

  img {
    width: inherit;
    height: 3rem;
  }

  img:hover {
    opacity: 0.8;
  }
`;

export const detailsStyle = css`
  display: flex;
  flex: 1;
  flex-direction: column;
  width: 100%;

  > h4 {
    color: var(--font-color);
  }

  > p {
    margin-top: 0.25em;
    color: var(--font-color-secondary);
    font-size: 0.875rem;
  }

  p:hover,
  h4:hover {
    opacity: 0.8;
  }
`;

export const hostStyle = css`
  a {
    &:hover,
    & h2:hover,
    & h3:hover {
      text-decoration: none;
    }
  }
`;

export const cardStyle = css`
  position: relative;
  display: flex;
  background-color: var(--bg-color);
  padding: 1.5rem;
  display: flex;
  flex-direction: row;
  align-items: center;
  height: 100%;
  justify-content: flex-start;
  box-shadow: rgba(0, 0, 0, 0.09) 0.3125rem 0.3125rem 0 -0.0625rem,
    0 0.25rem 0.5rem 0 rgba(0, 0, 0, 0.15);
  transition: all 0.6s cubic-bezier(0.165, 0.84, 0.44, 1);

  ${MQFablet} {
    padding: 2rem 1.5rem;
  }

  ${MQDesktop} {
    padding: 2rem 2.5rem;
  }

  &:hover {
    box-shadow: 0 0.375rem 0 -0.0625rem rgba(0, 0, 0, 0.1),
      0 0.25rem 0.5rem 0 rgba(0, 0, 0, 0.1);
    transform: translateY(-0.125rem);
  }
  &:active {
    box-shadow: 0rem 0.2rem 0 -0.025rem rgba(0, 0, 0, 0.2),
      0 0.0625rem 0.125rem 0 rgba(0, 0, 0, 0.2);
    transform: translateY(0);
  }

  &.vertical {
    flex-direction: column;
    justify-content: space-around;
    text-align: center;
    padding: 2.5rem;
  }
`;

export const externalLinkGraphic = css`
  position: absolute;
  top: 0.75rem;
  right: 0.75rem;
  width: 0.75rem;
`;
