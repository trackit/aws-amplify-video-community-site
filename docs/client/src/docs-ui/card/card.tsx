import {Component, h, Prop, Host} from "@stencil/core";
import {
  cardStyle,
  graphicStyle,
  detailsStyle,
  externalLinkGraphic,
  hostStyle,
} from "./card.style";
import {createVNode} from "../../utils/hyperscript";
import {pageContext} from "../../docs-ui/page/page.context";
import {SelectedFilters} from "../../docs-ui/page/page.types";
import {PLATFORM_FILTER_OPTIONS} from "../../utils/filter-data";

const mobilePlatforms = {
  ios: true,
  android: true,
};

const isMobileSelected = (selectedFilters?: SelectedFilters): boolean =>
  (selectedFilters?.platform && mobilePlatforms[selectedFilters.platform]) ||
  false;

const isPlatformSelected = (
  platform: typeof PLATFORM_FILTER_OPTIONS[number],
  selectedFilters?: SelectedFilters,
): boolean =>
  (selectedFilters?.platform && selectedFilters.platform === platform) || false;

@Component({tag: "docs-card", shadow: false})
export class DocsCard {
  /*** the global filter state */
  @Prop() readonly selectedFilters?: SelectedFilters;
  /*** add a different url when mobile selected */
  @Prop() readonly urlOverrideForMobileFilter?: string;
  /*** add a different url when mobile selected */
  @Prop() readonly urlOverrideForIOSFilter?: string;
  /*** add a different url when mobile selected */
  @Prop() readonly urlOverrideForAndroidFilter?: string;
  /*** add a different url when mobile selected */
  @Prop() readonly urlOverrideForJSFilter?: string;
  /*** if true, the thumbnail gets rendered to the left of the detail (not above) */
  @Prop() readonly vertical?: boolean;
  /*** url */
  @Prop() readonly url?: string;
  /*** query string parameters to attach to the link */
  @Prop() readonly QSPs: string = "";
  /*** link tag to use */
  @Prop() readonly containertag: string = "docs-internal-link";
  /*** whether or not to show external link graphic */
  @Prop() readonly external?: boolean;

  render() {
    const {
      vertical,
      urlOverrideForMobileFilter,
      urlOverrideForIOSFilter,
      urlOverrideForAndroidFilter,
      urlOverrideForJSFilter,
    } = this;

    let href = this.url;

    if (urlOverrideForMobileFilter && isMobileSelected(this.selectedFilters)) {
      href = urlOverrideForMobileFilter;
    }

    if (
      urlOverrideForIOSFilter &&
      isPlatformSelected("ios", this.selectedFilters)
    ) {
      href = urlOverrideForIOSFilter;
    }

    if (
      urlOverrideForAndroidFilter &&
      isPlatformSelected("android", this.selectedFilters)
    ) {
      href = urlOverrideForAndroidFilter;
    }

    if (
      urlOverrideForJSFilter &&
      isPlatformSelected("js", this.selectedFilters)
    ) {
      href = urlOverrideForJSFilter;
    }

    return (
      <Host class={hostStyle}>
        {createVNode(
          this.containertag,
          {
            href,
            QSPs: this.QSPs,
            ...(this.external ? {target: "_blank"} : {}),
          },
          <div class={{[cardStyle]: true, vertical: !!vertical}}>
            <div class={{[graphicStyle]: true, vertical: !!vertical}}>
              <slot name="graphic" />
            </div>
            <div class={detailsStyle}>
              <slot name="heading" />
              <slot name="description" />
            </div>
            {this.external && (
              <img
                class={externalLinkGraphic}
                src={`/assets/external-link-black.svg`}
                alt="External link"
              />
            )}
          </div>,
        )}
      </Host>
    );
  }
}

pageContext.injectProps(DocsCard, ["selectedFilters"]);
