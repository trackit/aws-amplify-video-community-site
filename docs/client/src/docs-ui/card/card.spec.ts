import {DocsCard} from "./card";
import {newSpecPage} from "@stencil/core/testing";

describe("amplify-card", () => {
  describe("Component logic", () => {
    let docsCard: DocsCard;
    beforeEach(() => (docsCard = new DocsCard()));

    it("should init `vertical` as `undefined`", () =>
      expect(docsCard.vertical).toBeUndefined());
  });

  describe("Render logic", () => {
    it("should render", async () => {
      expect(
        (
          await newSpecPage({
            components: [DocsCard],
            html: `<docs-card />`,
          })
        ).root,
      ).toMatchSnapshot();
    });
  });
});
