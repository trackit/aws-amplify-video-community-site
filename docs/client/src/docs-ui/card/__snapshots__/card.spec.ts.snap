// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`amplify-card Render logic should render 1`] = `
<docs-card class="css-1bcvumb">
  <!---->
  <docs-internal-link qsps="">
    <div class="css-fgu2xk">
      <div class="css-i36jdi"></div>
      <div class="css-16zu5pg">
      </div>
    </div>
  </docs-internal-link>
</docs-card>
`;
